doc/superbuild/source/developers/style_guide/Documentation.rst

```diff
@@ -127,28 +127,269 @@ All citations go into `doc/shared/sundials.bib`.
+   :nocontentsentry:
+   :noindexentry:
+
+   This function sets the value of Bar in a FooObject.
+
+   The default value for Bar is :math:`10`. An input value :math:`< 0` will
+   reset Bar to the default value.
+
+   :param foo_obj: the FooObject.
+   :param bar_value: the value of Bar.
+
+   :retval SUCCESS: if the value was successfully set.
+   :retval NULL_OBJ: if the ``foo_obj`` was ``NULL``.
+
+   .. versionadded:: 1.1.0
+
+   .. versionchanged:: 2.0.0
+
+      The type of p1 was changed from ``unsigned int`` to ``int``
+
+   .. note::
+
+      Utilizing this capability requires building with Bar enabled.
+
+   .. warning::
+
+      Setting values greater than 100 may degrade performance.
+
+   **Example usage:**
+
+   .. code-block:: C
+
+      /* Create the object */
+      void* foo_obj = CreateFooObject(p1, p2);
+      if (foo_obj == NULL) { return 1; }
+
+      /* Update the value of Bar */
+      int retval = FooSetBar(foo_obj, 50);
+      if (retval != 0) { return 1; }
+
+      /* Perform some actions */
+      ...
+
+   **Examples codes:**
+
+   * ``examples/package/subdir/pkg_foo_demo.c``
+
+
+User-Supplied Functions
+=======================
+
+Document user-supplied functions with the :external+sphinx:rst:dir:`c:type`
+directive. The directive is followed by the ``typedef`` for the function
+pointer. The description of the function type mirrors the style used for
+user-callable function with one exception. As :external+sphinx:rst:dir:`c:type`
+does not currently support the ``param``, ``retval``, and ``returns`` fields,
+these sections must be manually created. The style that follows is chosen to
+reflect that of ``param``, ``retval``, and ``returns`` fields as much as
+possible. Function parameters should listed under a boldface "Parameters:"
+section with the parameters in boldface and separated from their description by
+an en-dash. As user-supplied functions typically return a ``int``, but specific
+values are not required, a description of how the return value is interpreted
+should be given under a boldface "Returns:" section (skipping a line and
+indenting 2 spaces). If specific return values are required, these should
+documented similarly to the function parameters and listed under a boldface
+"Return values:" section. For example,
+
+.. code-block:: rst
+
+   .. c:type:: int (*FooFn)(param_type p1, param_type p2)
+
+      Brief description of what the user-provided function should do.
+
+      Additional information about the function and its usage.
+
+      **Parameters:**
+
+      * **p1** -- description of the first parameter.
+      * **p2** -- description of the second parameter.
+
+      **Returns:**
+
+        A :c:type:`FooFn` function should return 0 if successful, a positive
+        value if a recoverable error occurred, or a negative value if an
+        unrecoverable error occurred.
+
+.. code-block:: rst
+
+   .. c:type:: int (*FooFn)(param_type p1, param_type p2)
```

Comment:
Since the previous example defined `FooFn`, perhaps this should be a different typedef?
```suggestion
   .. c:type:: int (*BarFn)(param_type p1, param_type p2)
```